doc/design/api.md

```diff
@@ -0,0 +1,87 @@
+   
+After the invocation of the backward algorithm, we should update model parameters using the gradients and parameters like learning rate.  This step might involve communications with the parameter server in the case of distributed training.  This complexity motivates us to separate the trainer into two concepts:
+   
+1. *gradient machine*, which computes and maintains layer outputs and gradients, and
+      
+1. *updater*, which encapsulates the updating algorithm.
+   
+It seems that *cost function* is a property of *gradient machine*?
+   
+### Data
+
+Models are trained using data sets.  We hope to provide a set of utility data sets encapsulated in Python packages like `paddle.data.amazon.product_review` and `paddle.data.wikipedia.articles`.  A reasonable idea might be that in each of these packages, we provide a `new` function that returns a reader object or a Python iterator.  And the *reader* has a read method `read`, which, once called, returns a minibatch and a flag indicating if it reaches the end of a data set.  For online learning, the flag would always be False.  In this way, we don't have to have two concepts: *pass* and *iteration*; instead, we need only the latter.
+
+## Examples
+
+### A Simple Network
+
+```python
+gm = paddle.gradient_machine.new(model)  # gm uses default cost function.
+ud = paddle.updater.new_simple()  # A simple updater doesn't work with parameter servers.
+rd = paddle.data.amazon.product_review.new()
+mt = paddle.metric.new_acc()
+
+for mb, pass_end in rd.read():
+    gm.feed(mb)
+    ud.update(gm)
+    mt.append(gm, mb) # compute and record the accuracy of this minibatch.
+    if pass_end:
+        log(mt.flash()) # output aggregated accuracy records of this pass and reset mt.
+```
+
+In this example, `GradientMachine.feed` is a convenience that calls `GradientMachine.forward` and `GradientMachine.backward`.
+
+### A GAN Example
+
+```python
+input_gen = paddle.layer.input(...)
+hidden_gen = paddle.layer.fc(input_gen, ...)
+output_gen = paddle.layer.fc(hidden_gen, ...)
+
+# gm_gen maintains layer outputs and gradients of model gen.
+gm_gen = paddle.gradient_machine.new(output_gen)
+
+input_dis = paddle.layer.input(...)
+hidden_dis = paddle.layer.fc(intput_dis, ...)
+output_dis = paddle.layer.softmax(hidden_dis, ...)
+
+# gm_dis maintains layer outputs and gradients of model dis.
+gm_dis = paddle.gradient_machine.new(output_dis)
+
+# gm_comp maintains layer outputs and gradients of both gen and dis.
+gm_comp = paddle.gradient_machine.compose(output_gen, output_dis)
+
+ud = paddle.updater.new_simple()
+
+rd = paddle.data.mnist.new()
+
+for mb, pass_end in rd.read():
+    fake = gm_gen.forward(mb)
+    fake_label = paddle.input.new(False, ...)
+    real_label = paddle.input.new(True, ...)
+    gm_dis.feed([fake, fake_label])
+    gm_dis.feed([mb, real_label])
+    ud.update(gm_dis)
+    
+    gm_comp.feed([mb, real_label])
```

Comment:
Since updater will only need to update gradients of `output_gen`'s predecessors'. One possible optimization here is to let gradient machine know this information. So it does not need to save any unrelated activation and gradients. E.g.,
```
gm_comp.feed([mb, real_label], output_gen)
ud.update(gm_comp, output_gen)
```

Maybe I don't have enough context, but from a first glance, I feel maybe it's easier that gradient machine be stateless (current design saves gradients inside gradient machine). E.g.,
```
// pseudo code
type gradientMachine
type gradient
type updater

var gm_comp gradientMachine
var ud updater
var gradient g = gm_comp.feed([mb, real_label], output_gen)
ud.update(g)
```